freebsd: netdata is killed by signal 10

@vlvkobal I tried to install the latest version of netdata at https://freebsd.my-netdata.io.

It starts and is immediately killed by signal 10 (SIGBUS).

I compiled netdata with `CFLAGS="-O1 -ggdb -Wall -Wextra"` and I got warnings:

```
freebsd_sysctl.c: In function 'do_vm_vmtotal':
freebsd_sysctl.c:124:44: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_vmtotal(int update_every, usec_t dt) {
                                            ^
freebsd_sysctl.c: In function 'do_kern_cp_time':
freebsd_sysctl.c:242:46: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_kern_cp_time(int update_every, usec_t dt) {
                                              ^
freebsd_sysctl.c: In function 'do_kern_cp_times':
freebsd_sysctl.c:299:47: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_kern_cp_times(int update_every, usec_t dt) {
                                               ^
freebsd_sysctl.c: In function 'do_hw_intcnt':
freebsd_sysctl.c:399:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             for (i = 0; i < nintr; i++)
                           ^
freebsd_sysctl.c:464:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
                 for (i = 0; i < nintr; i++) {
                               ^
freebsd_sysctl.c:376:43: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_hw_intcnt(int update_every, usec_t dt) {
                                           ^
freebsd_sysctl.c: In function 'do_vm_stats_sys_v_intr':
freebsd_sysctl.c:484:53: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_stats_sys_v_intr(int update_every, usec_t dt) {
                                                     ^
freebsd_sysctl.c: In function 'do_vm_stats_sys_v_soft':
freebsd_sysctl.c:526:53: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_stats_sys_v_soft(int update_every, usec_t dt) {
                                                     ^
freebsd_sysctl.c: In function 'do_vm_stats_sys_v_swtch':
freebsd_sysctl.c:568:54: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_stats_sys_v_swtch(int update_every, usec_t dt) {
                                                      ^
freebsd_sysctl.c: In function 'do_vm_stats_sys_v_forks':
freebsd_sysctl.c:610:54: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_stats_sys_v_forks(int update_every, usec_t dt) {
                                                      ^
freebsd_sysctl.c: In function 'do_vm_swap_info':
freebsd_sysctl.c:654:46: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_swap_info(int update_every, usec_t dt) {
                                              ^
freebsd_sysctl.c: In function 'do_system_ram':
freebsd_sysctl.c:729:44: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_system_ram(int update_every, usec_t dt) {
                                            ^
freebsd_sysctl.c: In function 'do_vm_stats_sys_v_swappgs':
freebsd_sysctl.c:796:56: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_stats_sys_v_swappgs(int update_every, usec_t dt) {
                                                        ^
freebsd_sysctl.c: In function 'do_vm_stats_sys_v_pgfaults':
freebsd_sysctl.c:841:57: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_vm_stats_sys_v_pgfaults(int update_every, usec_t dt) {
                                                         ^
freebsd_sysctl.c: In function 'do_kern_ipc_sem':
freebsd_sysctl.c:899:25: warning: unused parameter 'update_every' [-Wunused-parameter]
 int do_kern_ipc_sem(int update_every, usec_t dt) {
                         ^
freebsd_sysctl.c:899:46: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_kern_ipc_sem(int update_every, usec_t dt) {
                                              ^
freebsd_sysctl.c: In function 'do_kern_ipc_shm':
freebsd_sysctl.c:1012:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             for (i = 0; i < ipc_shm.shmmni; i++) {
                           ^
freebsd_sysctl.c:986:25: warning: unused parameter 'update_every' [-Wunused-parameter]
 int do_kern_ipc_shm(int update_every, usec_t dt) {
                         ^
freebsd_sysctl.c:986:46: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_kern_ipc_shm(int update_every, usec_t dt) {
                                              ^
freebsd_sysctl.c: In function 'do_kern_ipc_msq':
freebsd_sysctl.c:1074:25: warning: unused parameter 'update_every' [-Wunused-parameter]
 int do_kern_ipc_msq(int update_every, usec_t dt) {
                         ^
freebsd_sysctl.c:1074:46: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_kern_ipc_msq(int update_every, usec_t dt) {
                                              ^
freebsd_sysctl.c: In function 'do_uptime':
freebsd_sysctl.c:1192:40: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_uptime(int update_every, usec_t dt) {
                                        ^
freebsd_sysctl.c: In function 'do_net_isr':
freebsd_sysctl.c:1281:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
             for (i = 0; i < num_netisr_workstreams; i++) {
                           ^
freebsd_sysctl.c:1282:31: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
                 for (n = 0; n < num_netisr_works; n++) {
                               ^
freebsd_sysctl.c:1228:41: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_isr(int update_every, usec_t dt) {
                                         ^
freebsd_sysctl.c: In function 'do_net_inet_tcp_states':
freebsd_sysctl.c:1395:53: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_inet_tcp_states(int update_every, usec_t dt) {
                                                     ^
freebsd_sysctl.c: In function 'do_net_inet_tcp_stats':
freebsd_sysctl.c:1438:52: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_inet_tcp_stats(int update_every, usec_t dt) {
                                                    ^
freebsd_sysctl.c: In function 'do_net_inet_udp_stats':
freebsd_sysctl.c:1732:52: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_inet_udp_stats(int update_every, usec_t dt) {
                                                    ^
freebsd_sysctl.c: In function 'do_net_inet_icmp_stats':
freebsd_sysctl.c:1832:53: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_inet_icmp_stats(int update_every, usec_t dt) {
                                                     ^
freebsd_sysctl.c: In function 'do_net_inet_ip_stats':
freebsd_sysctl.c:1957:51: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_inet_ip_stats(int update_every, usec_t dt) {
                                                   ^
freebsd_sysctl.c: In function 'do_net_inet6_ip6_stats':
freebsd_sysctl.c:2139:53: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_inet6_ip6_stats(int update_every, usec_t dt) {
                                                     ^
freebsd_sysctl.c: In function 'do_net_inet6_icmp6_stats':
freebsd_sysctl.c:2349:55: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_net_inet6_icmp6_stats(int update_every, usec_t dt) {
                                                       ^
freebsd_sysctl.c: In function 'do_getmntinfo':
freebsd_sysctl.c:2717:44: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_getmntinfo(int update_every, usec_t dt) {
                                            ^
freebsd_sysctl.c: In function 'do_getifaddrs':
freebsd_sysctl.c:2922:44: warning: unused parameter 'dt' [-Wunused-parameter]
 int do_getifaddrs(int update_every, usec_t dt) {
                                            ^
freebsd_sysctl.c: In function 'do_kern_devstat':
freebsd_sysctl.c:3469:25: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
                         if (dstat[i].device_type & DEVSTAT_TYPE_PASS == DEVSTAT_TYPE_PASS)
                         ^
freebsd_sysctl.c:3532:37: warning: missing initializer for field 'bytes_write' of 'struct prev_dstat' [-Wmissing-field-initializers]
                                     .prev_dstat.bytes_write       = dstat[i].bytes[DEVSTAT_WRITE],
                                     ^
freebsd_sysctl.c:3404:42: note: 'bytes_write' declared here
                         collected_number bytes_write;
                                          ^
freebsd_sysctl.c:3533:37: warning: missing initializer for field 'operations_read' of 'struct prev_dstat' [-Wmissing-field-initializers]
                                     .prev_dstat.operations_read   = dstat[i].operations[DEVSTAT_READ],
                                     ^
freebsd_sysctl.c:3405:42: note: 'operations_read' declared here
                         collected_number operations_read;
                                          ^
freebsd_sysctl.c:3534:37: warning: missing initializer for field 'operations_write' of 'struct prev_dstat' [-Wmissing-field-initializers]
                                     .prev_dstat.operations_write  = dstat[i].operations[DEVSTAT_WRITE],
                                     ^
freebsd_sysctl.c:3406:42: note: 'operations_write' declared here
                         collected_number operations_write;
                                          ^
freebsd_sysctl.c:3535:37: warning: missing initializer for field 'duration_read_ms' of 'struct prev_dstat' [-Wmissing-field-initializers]
                                     .prev_dstat.duration_read_ms  = dstat[i].duration[DEVSTAT_READ].sec * 1000
                                     ^
freebsd_sysctl.c:3407:42: note: 'duration_read_ms' declared here
                         collected_number duration_read_ms;
                                          ^
freebsd_sysctl.c:3537:37: warning: missing initializer for field 'duration_write_ms' of 'struct prev_dstat' [-Wmissing-field-initializers]
                                     .prev_dstat.duration_write_ms = dstat[i].duration[DEVSTAT_WRITE].sec * 1000
                                     ^
freebsd_sysctl.c:3408:42: note: 'duration_write_ms' declared here
                         collected_number duration_write_ms;
                                          ^
freebsd_sysctl.c:3539:37: warning: missing initializer for field 'busy_time_ms' of 'struct prev_dstat' [-Wmissing-field-initializers]
                                     .prev_dstat.busy_time_ms      = dstat[i].busy_time.sec * 1000
                                     ^
freebsd_sysctl.c:3409:42: note: 'busy_time_ms' declared here
                         collected_number busy_time_ms;
                                          ^
```

There are more warnings at other places too.

Then I run it under valgrind with this:

```
valgrind /opt/netdata/usr/sbin/netdata -D
```

and got this error twice:

```
==19883== Conditional jump or move depends on uninitialised value(s)
==19883==    at 0x42E436: do_net_isr (freebsd_sysctl.c:1353)
==19883==    by 0x453B8D: freebsd_main (plugin_freebsd.c:105)
==19883==    by 0x547AB54: ??? (in /lib/libthr.so.3)
```

Also, hitting control-C on valgrind, netdata core dumps:

```
==19883== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==19883==  General Protection Fault
==19883==    at 0x578713C: ??? (in /lib/libc.so.7)
==19883==    by 0x57DAAC5: vsprintf_l (in /lib/libc.so.7)
==19883==    by 0x57D8AB0: sprintf_l (in /lib/libc.so.7)
==19883==    by 0x5781C56: ??? (in /lib/libc.so.7)
==19883==    by 0x5780F63: ??? (in /lib/libc.so.7)
==19883==    by 0x57806BA: strftime_l (in /lib/libc.so.7)
==19883==    by 0x4306E4: log_date (log.c:219)
==19883==    by 0x430F15: error_int (log.c:306)
==19883==    by 0x43193C: sig_handler_exit (daemon.c:10)
==19883==    by 0x548078E: ??? (in /lib/libthr.so.3)
==19883==    by 0x547FD6E: ??? (in /lib/libthr.so.3)
==19883==    by 0x380C0657: ??? (in /usr/local/lib/valgrind/memcheck-amd64-freebsd)
```

I suggest to always work with compiler warnings enabled, like this:

```
CFLAGS="-O1 -ggdb -Wall -Wextra -fstack-protector-all -DNETDATA_INTERNAL_CHECKS=1" ./netdata-installer.sh --install /opt
```

--- a/src/freebsd_sysctl.c
+++ b/src/freebsd_sysctl.c
@@ -330,7 +330,7 @@ int do_kern_cp_times(int update_every, usec_t dt) {
 
             if(unlikely(number_of_cpus > old_number_of_cpus)) {
                 all_cpu_charts = reallocz(all_cpu_charts, sizeof(struct cpu_chart) * number_of_cpus);
-                memset(&all_cpu_charts[old_number_of_cpus], 0, sizeof(struct softnet_chart) * (number_of_cpus - old_number_of_cpus));
+                memset(&all_cpu_charts[old_number_of_cpus], 0, sizeof(struct cpu_chart) * (number_of_cpus - old_number_of_cpus));
                 old_number_of_cpus = number_of_cpus;
             }
 

--- a/src/freebsd_sysctl.c
+++ b/src/freebsd_sysctl.c
@@ -326,8 +326,13 @@ int do_kern_cp_times(int update_every, usec_t dt) {
                 RRDDIM *rd_interrupt;
                 RRDDIM *rd_idle;
             } *all_cpu_charts = NULL;
+            static int old_number_of_cpus = 0;
 
-            all_cpu_charts = reallocz(all_cpu_charts, sizeof(struct cpu_chart) * number_of_cpus);
+            if(unlikely(number_of_cpus > old_number_of_cpus)) {
+                all_cpu_charts = reallocz(all_cpu_charts, sizeof(struct cpu_chart) * number_of_cpus);
+                memset(&all_cpu_charts[old_number_of_cpus], 0, sizeof(struct softnet_chart) * (number_of_cpus - old_number_of_cpus));
+                old_number_of_cpus = number_of_cpus;
+            }
 
             for (i = 0; i < number_of_cpus; i++) {
                 if (unlikely(!all_cpu_charts[i].st)) {
@@ -1277,7 +1282,7 @@ int do_net_isr(int update_every, usec_t dt) {
             return 1;
         } else {
             netisr_stats = reallocz(netisr_stats, (number_of_cpus + 1) * sizeof(struct netisr_stats));
-            bzero(netisr_stats, (number_of_cpus + 1) * sizeof(struct netisr_stats));
+            memset(netisr_stats, 0, (number_of_cpus + 1) * sizeof(struct netisr_stats));
             for (i = 0; i < num_netisr_workstreams; i++) {
                 for (n = 0; n < num_netisr_works; n++) {
                     if (netisr_workstream[i].snws_wsid == netisr_work[n].snw_wsid) {
@@ -1344,8 +1349,13 @@ int do_net_isr(int update_every, usec_t dt) {
             RRDDIM *rd_qdrops;
             RRDDIM *rd_queued;
         } *all_softnet_charts = NULL;
+        static int old_number_of_cpus = 0;
 
-        all_softnet_charts = reallocz(all_softnet_charts, sizeof(struct softnet_chart) * number_of_cpus);
+        if(unlikely(number_of_cpus > old_number_of_cpus)) {
+            all_softnet_charts = reallocz(all_softnet_charts, sizeof(struct softnet_chart) * number_of_cpus);
+            memset(&all_softnet_charts[old_number_of_cpus], 0, sizeof(struct softnet_chart) * (number_of_cpus - old_number_of_cpus));
+            old_number_of_cpus = number_of_cpus;
+        }
 
         for (i = 0; i < number_of_cpus ;i++) {
             snprintfz(all_softnet_charts[i].netisr_cpuid, MAX_INT_DIGITS + 17, "cpu%d_softnet_stat", i);

